eltwise.hpp

```diff
@@ -0,0 +1,147 @@
+ * StreamingEltwise
+ * 
+ * \tparam TIn0 Datatype of the first input to the eltwise function
+ * \tparam TIn1 Datatype of the second input to the eltwise function
+ * \tparam TOut Datatype of the output generated by the eltwise function
+ *
+ */
+template<typename TIn0, typename TIn1, typename TOut>
+class EltwiseFunction {
+public:
+/*!
+ * \brief eltwise_op: computes the eltwise algorithm on a pair of elements
+ *
+ * \param input0 First input value to be used in the eltwise function 
+ * \param input1 Second input value to be used in the eltwise function 
+*/
+    TOut eltwise_op(TIn0 const &input0, TIn1 const &input1) const;
+};
+
+template<typename TIn0, typename TIn1, typename TOut>
+class AddEltwiseFunction : public EltwiseFunction<TIn0, TIn1, TOut> {
+public:
+    TOut eltwise_op(TIn0 const &input0, TIn1 const &input1) const {
+#pragma HLS inline
+        return (TOut)(input0 + input1);
+    }
+};
+
+template<typename TIn0, typename TIn1, typename TOut>
+class SubEltwiseFunction : public EltwiseFunction<TIn0, TIn1, TOut> {
+public:
+    TOut eltwise_op(TIn0 const &input0, TIn1 const &input1) const {
+#pragma HLS inline
+        return (TOut)(input0 - input1);
+    }
+};
+
+template<typename TIn0, typename TIn1, typename TOut>
+class AbsDiffEltwiseFunction : public EltwiseFunction<TIn0, TIn1, TOut> {
+public:
+    TOut eltwise_op(TIn0 const &input0, TIn1 const &input1) const {
+#pragma HLS inline
+        return (TOut)((input0 > input1) ? (input0-input1) : (input1-input0));
+    }
+};
+
+/**
+ * \brief StreamingEltwise function
+ *
+ * The function performs a generic eltwise function on two streams and
+ * produces an output stream.
+ *
+ * \tparam Channels   Number of channels for eltwise operation
+ * \tparam PE         Number of channels for eltwise operation computed in parallel
+ * \tparam N          Total number of elements to process
+ * \tparam SliceIn0   Data slicer for input 0 type
+ * \tparam SliceIn1   Data slicer for input 1 type
+ * \tparam SliceOut   Data slicer for output type
+ * \tparam TStrmIn0   Type of the input 0 stream - safely deducible from the paramaters
+ * \tparam TStrmIn1   Type of the input 1 stream - safely deducible from the paramaters
+ * \tparam TStrmOut   Type of the output - safely deducible from the paramaters
+ * \tparam TFxn       Type of the function class (e.g. Max, Avg, Sum) - safely deducible from the paramaters
+ *
+ * \param in          Input stream
+ * \param out         Output stream
+ * \param function    Function to apply, derived from EltwiseFunction
+ */
+template<
+  unsigned Channels, unsigned PE, unsigned N,
+  typename SliceIn0,typename SliceIn1, typename SliceOut,
+  typename TStrmIn0, typename TStrmIn1, typename TStrmOut,
+  typename TFxn
+>
+void StreamingEltwise(hls::stream<TStrmIn0> &in0,
+    hls::stream<TStrmIn1> &in1,
+    hls::stream<TStrmOut> &out,
+    TFxn const &function) {
```

Comment:
Use a [forwarding reference](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references) to become compatible with lambdas and other functors.